Mathlib/Tactic/Pushneg.lean

```diff
@@ -0,0 +1,119 @@
+    let (eNew, eqProof) ← pushNegation e
+    if eqProof.isAppOf ``Eq.refl
+    then return (expr, ← mkEqRefl expr)
+    else return (←mkLambdaFVars #[x] eNew, ← mkFunExt (←mkLambdaFVars #[x] eqProof))
+| Expr.app e₁ e₂ _ =>
+   match expr.not? with
+   | some expr' =>
+      match expr'.getAppFnArgs with
+      | (`Not, #[e]) =>
+        let (eNew, eqProof) ← pushNegation e
+        let eqProof ← mkAppOptM ``not_not_eq #[none, none, eqProof]
+        return (eNew, eqProof)
+      | (`And, #[p, q]) =>
+        let (p, eqProof₁) ← pushNegation p
+        let (q, eqProof₂) ← pushNegation (mkNot q)
+        let eNew := mkForall `_ default p q
+        let eqProof ← mkAppOptM ``not_and_eq #[none, none,none, none, eqProof₁, eqProof₂]
+        return (eNew, eqProof)
+      | (`Or, #[p, q]) =>
+        let (p, eqProof₁) ← pushNegation (mkNot p)
+        let (q, eqProof₂) ← pushNegation (mkNot q)
+        let eNew := mkAnd p q
+        let eqProof ← mkAppOptM ``not_or_eq #[none, none,none, none, eqProof₁, eqProof₂]
+        return (eNew, eqProof)
+      | (`Exists, #[t, e]) =>
+        binderTelescope1 e fun x e => do
+          let (eNew, eqProof) ← pushNegation (mkNot e)
+          let eNew ← mkForallFVars #[x] eNew
+          let eqProof ← mkLambdaFVars #[x] eqProof
+          let eqProof ← mkAppOptM ``not_exists_eq #[none, none, none, eqProof]
+          return (eNew, eqProof)
+      | _ => match expr' with
+        | Expr.forallE _ t _ _ =>
+          binderTelescope1 expr' fun x e => do
+            let (eNew, eqProof) ← pushNegation (mkNot e)
+            let eNew ← mkLambdaFVars #[x] eNew
+            let level ← getLevel (←inferType eNew)
+            let eNew := mkAppN (mkConst ``Exists [level]) #[t, eNew]
+            let eqProof ← mkLambdaFVars #[x] eqProof
+            let eqProof ← mkAppOptM ``not_forall_eq #[none, none, none, eqProof]
+            return (eNew, eqProof)
+        | _ =>
+          let (eNew, eqProof) ← pushNegation expr'
+          let eqProof ← mkCongr (←mkEqRefl (mkConst `Not)) eqProof
+          return (mkNot eNew, eqProof)
+   | _ =>
+      let (e₁', eqProof₁) ← pushNegation e₁
+      let (e₂', eqProof₂) ← pushNegation e₂
+      let eqProof ← mkCongr eqProof₁ eqProof₂
+      return (mkApp e₁' e₂', eqProof)
+| _ => return (expr, ←mkEqRefl expr)
+
+
+def pushnegTarget : TacticM Unit := do
+let target ← getMainTarget
+let (eNew, eqProof) ← pushNegation target
+if !eqProof.isAppOf ``Eq.refl
+then
+  let mvarId' ← replaceTargetEq (← getMainGoal) eNew eqProof
+  replaceMainGoal [mvarId']
+else
+  throwError "Pushneg couldn't find a negation to push"
+
+def pushNegLocalDecl (fvarId : FVarId): TacticM Unit := do
+let target ← getLocalDecl fvarId
+let (eNew, eqProof) ← pushNegation target.type
+if ! eqProof.isAppOf ``Eq.refl
+then
+  let mvarId' ← replaceLocalDecl (← getMainGoal) fvarId  eNew eqProof
+  replaceMainGoal [mvarId'.mvarId]
+else
+  throwError "Pushneg couldn't find a negation to push"
```

Comment:
```suggestion
  throwError "push_neg couldn't find a negation to push"
```

BTW, I don't think that was the behavior in Lean 3.  AFAICT push_neg doesn't fail if it can't do anything.